LanguageExtension downcast resolution

Tanish-Ranjan opened the pull request:
This PR removes the usage of a class loader for downcasting `LanguageExtesnion` to `JavaExtension` and `ScalaExtension`. Previously, the client JVM would attempt to determine the appropriate sub-interface type using a class loader.

### Changes:

- Introduced new methods in the `LanguageExtension` interface:
  - `isJavaExtension()` and `isScalaExtension()`: Checks if the current implementation is a specific sub-interface type.
  - `getAsJavaExtension()` and `getAsScalaExtension()`: Returns the instance as the specific sub-interface type if applicable.
- Moved the conversion logic from the client JVM to the Gradle daemon JVM, where the actual implementation of the `LanguageExtension` interface resides. This ensures correct type information is available for downcasting.
- Updated respective test cases to reflect the new approach.

Tanish-Ranjan committed:
Refactor LanguageExtension interface: dependency injection for sub-interfaces
This commit refactors the LanguageExtension interface to eliminate the need for a class loader to convert it to its sub-interfaces JavaExtension and ScalaExtension.
Changes:
- Removed usage of class loader for conversion.
- Introduced methods within LanguageExtension to check if the current implementation is Java or Scala Extension and return specific instance if applicable.

Tanish-Ranjan committed:
Update test cases for LanguageExtension refactoring
This commit updates unit tests to reflect the changes in the LanguageExtension interface.
Changes:
- Mocks for JavaExtension and ScalaExtension are updated to return correct values for the newly introduced methods.
These updates ensure that tests continue to verify the expected behavior of LanguageExtension.

jdneo commented on model/build.gradle:
```diff
@@ -1,3 +1,7 @@
 }
+
+dependencies {
+  implementation 'org.gradle:gradle-tooling-api:8.7'
```
Let's use 8.8, which is the latest

Tanish-Ranjan replied:
Updated 👍

jdneo commented on model/src/main/java/com/microsoft/java/bs/gradle/model/LanguageExtension.java:
```diff
@@ -3,17 +3,25 @@
 
 package com.microsoft.java.bs.gradle.model;
 
+import org.gradle.internal.impldep.javax.annotation.Nullable;
```
Hey, @hegyibalint do you know if this is fine here? Because I see the Nullable class comes from gradle internal package.

hegyibalint replied:
I would rather not, as I assume IDE/QA tooling wouldn't pick up this re-packaged class.
If you want to have the annotations, you can depend on it directly just like this:

https://github.com/gradle/gradle/blob/d40c6121608cad99eb818fa698a4daa3eac99aee/build-logic/dependency-modules/src/main/kotlin/gradlebuild/modules/extension/ExternalModulesExtension.kt#L108

jdneo replied:
I see, thanks @hegyibalint 

@Tanish-Ranjan I think this can be another separate engineer work to enable null check to this project. For now, I think we can just remove those annotation in this PR, and make it clear that it may return null in the javadoc.

Tanish-Ranjan replied:
Done ✅

jdneo reviewed (commented):
LGTM. Just some minor issues.

Tanish-Ranjan committed:
Update tooling api version to 8.8 in model module.

Tanish-Ranjan committed:
Update LanguageExtension
Changes:
- removed @Nullable annotation which came from gradle's internal package.
- added JavaDoc

Arthurm1 commented on model/build.gradle:
```diff
@@ -1,3 +1,7 @@
+
+dependencies {
+  implementation 'org.gradle:gradle-tooling-api:8.8'
+}
```
Is this needed now `@Nullable` is gone?

jdneo replied:
Yes it's not needed. I'm just about to append a commit to remove it

jdneo committed:
Remove the gradle tooling api from model

jdneo commented:
Thank you @Tanish-Ranjan for your contribution! You can continue to work on #154 now.
